src/main/java/com/direwolf20/buildinggadgets/common/tiles/EffectBlockTileEntity.java

```diff
@@ -0,0 +1,123 @@
+package com.direwolf20.buildinggadgets.common.tiles;
+
+import com.direwolf20.buildinggadgets.api.Registries;
+import com.direwolf20.buildinggadgets.api.abstraction.BlockData;
+import com.direwolf20.buildinggadgets.common.blocks.EffectBlock.Mode;
+import com.direwolf20.buildinggadgets.common.registry.objects.BGBlocks.BGTileEntities;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.network.NetworkManager;
+import net.minecraft.network.play.server.SUpdateTileEntityPacket;
+import net.minecraft.tileentity.ITickableTileEntity;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.world.World;
+
+public class EffectBlockTileEntity extends TileEntity implements ITickableTileEntity {
+
+    /**
+     * Even though this is called "rendered", is will be used for replacement under normal conditions.
+     */
+    private BlockData renderedBlock;
+    /**
+     * A copy of the target block, used for inheriting data for {@link Mode#REPLACE}
+     */
+    private BlockData sourceBlock;
+
+    private Mode mode = null;
+    private boolean usePaste;
+
+    private int ticks;
+
+    public EffectBlockTileEntity() {
+        super(BGTileEntities.EFFECT_BLOCK_TYPE);
+    }
+
+    public void initializeData(World world, BlockData replacementBlock, Mode mode, boolean usePaste) {
+        // Minecraft will reuse a tile entity object at a location where the block got removed, but the modification is still buffered, and the block got restored again
+        // If we don't reset this here, the 2nd phase of REPLACE will simply finish immediately because the tile entity object is reused
+        this.ticks = 0;
+        // Again we don't check if the data has been set or not because there is a chance that this tile object gets reused
+
+        BlockData currentBlock = Registries.TileEntityData.createBlockData(world, pos);
+        TileEntity te = world.getTileEntity(pos);
+        this.sourceBlock = replacementBlock;
+
+        this.mode = mode;
+        this.usePaste = usePaste;
+
+        if (mode == Mode.REPLACE)
```

Comment:
Could you move the createBlockData step here, and avoid it when not needed? Creating a BlockData is a linear time operation in the amount of DataProviderFactories.